modules/renter/contractor/contractor.go

```diff
@@ -81,41 +76,76 @@ func (c *Contractor) SetAllowance(a modules.Allowance) error {
 		return errors.New("renew window must be less than period")
 	}
 
-	err := c.formContracts(a)
+	// check that allowance is sufficient to store at least one sector
+	numSectors, err := maxSectors(a, c.hdb)
 	if err != nil {
 		return err
+	} else if numSectors == 0 {
+		return errInsufficientAllowance
 	}
 
-	// Set the allowance.
+	// prevent ProcessConsensusChange from renewing contracts until we have
+	// finished
+	c.contractLock.Lock()
+	defer c.contractLock.Unlock()
+
 	c.mu.Lock()
-	c.allowance = a
-	err = c.saveSync()
+	endHeight := c.blockHeight + a.Period
+
+	// check if allowance has different period or funds; if so, we should
+	// renew existing contracts with the new allowance.
+	// TODO: compare numSectors instead of allowance.Funds?
+	var renewSet []modules.RenterContract
+	if a.Period != c.allowance.Period || a.Funds.Cmp(c.allowance.Funds) != 0 {
+		for _, contract := range c.contracts {
+			renewSet = append(renewSet, contract)
+		}
+		// TODO: dangerous -- figure out a safer approach later
+		c.contracts = map[types.FileContractID]modules.RenterContract{}
```

Comment:
Hmm, I don't think you should be deleting contracts until the renew for that contract has completed. Ideally upon restart it would pick up where it left off, including renewing to a homogeneous height (same height as the previous ones were already renewed to)

Definitely lots of edge cases here, more than I was expecting.